### PR TITLE
Performance booster

### PR DESCRIPTION
**Pull request by jwinzer**
Voting performance has been dramatically improved.

### CODE REVIEW

File: openslides_votecollector/api.py [hunk 1]
```diff
@@ -41,25 +41,19 @@ def __str__(self):
 
 def get_server():
     """
-    Gets a server proxy object and tests the connection.
+    Gets a server proxy object.
     """
     try:
-        server = ServerProxy(config['votecollector_uri'])
-        # TODO: reduce timeout
+        return ServerProxy(config['votecollector_uri'])
```
Comment (normanjaeckel) [inline openslides_votecollector/api.py]: Coding style proposal: Do not use return inside the try block. The last line of each function should be the return line:

```python
try:
    server = ServerProxy(config['votecollector_uri'])
except TypeError:
    raise VoteCollectorError(_('Server not found.'))
return server
```

File: openslides_votecollector/api.py [hunk 2]
```diff
@@ -41,25 +41,19 @@ def __str__(self):
...
-        server.voteCollector.getDeviceStatus()
-    except:
-        raise VoteCollectorError(_('No connection to VoteCollector.'))
-    else:
-        return server
-
 
 def get_keypads():
-    keypads = Keypad.objects.exclude(user__is_present=False).values_list(
+    # FIXME: Wohnbau
```
Comment (normanjaeckel) [inline openslides_votecollector/api.py]: Remove these comments.

File: openslides_votecollector/api.py [hunk 3]
```diff
@@ -75,28 +69,37 @@ def get_keypads():
 
 def get_device_status():
     server = get_server()
-    return server.voteCollector.getDeviceStatus()
+    try:
+        return server.voteCollector.getDeviceStatus()
```
Comment (normanjaeckel) [inline openslides_votecollector/api.py]: Here the same.

File: openslides_votecollector/api.py [hunk 4]
```diff
@@ -75,28 +69,37 @@ def get_keypads():
...
+        if count < 0:
+            raise VoteCollectorError(nr=count)
 
-    count = server.voteCollector.startVoting()
-    if count < 0:
-        raise VoteCollectorError(nr=count)
+        count = server.voteCollector.startVoting()
+        if count < 0:
+            raise VoteCollectorError(nr=count)
+    except:
```
Comment (normanjaeckel) [inline openslides_votecollector/api.py]: Do not use global except statement. Please specify the expected exception.

File: openslides_votecollector/api.py [hunk 5]
```diff
@@ -75,28 +69,37 @@ def get_keypads():
...
 def start_voting(mode, options, callback_url):
     server = get_server()
     keypads = get_keypads()
 
     ext_mode = options + ';' + callback_url if options else callback_url
-    count = server.voteCollector.prepareVoting(mode + '-' + ext_mode, 0, 0, list(keypads))
-    if count < 0:
-        raise VoteCollectorError(nr=count)
+    try:
+        count = server.voteCollector.prepareVoting(mode + '-' + ext_mode, 0, 0, list(keypads))
```
Comment (normanjaeckel) [inline openslides_votecollector/api.py]: Please try to minimize the length of the try block. Try to use only one statement inside the try block. This should be the statement that may raise the exception.

File: openslides_votecollector/api.py [hunk 6]
```diff
@@ -105,13 +108,18 @@ def get_voting_status():
     Returns voting status as a list: [elapsed_seconds, votes_received]
     """
     server = get_server()
-    status = server.voteCollector.getVotingStatus()
-    return status
+    try:
+        return server.voteCollector.getVotingStatus()
```
Comment (normanjaeckel) [inline openslides_votecollector/api.py]: Here the same.

File: openslides_votecollector/api.py [hunk 7]
```diff
@@ -105,13 +108,18 @@ def get_voting_status():
...
 
 
 def get_voting_result():
     """
     Returns the voting result as a list.
     """
     server = get_server()
-    return server.voteCollector.getVotingResult()
+    try:
+        return server.voteCollector.getVotingResult()
```
Comment (normanjaeckel) [inline openslides_votecollector/api.py]: Here the same.

File: openslides_votecollector/seating_plan.py [hunk 8]
```diff
@@ -5,6 +5,28 @@ def setup_default_plan():
     """
     Adds a default seating plan.
     """
+    limit = 200
```
Comment (normanjaeckel) [inline openslides_votecollector/seating_plan.py]: I do not understand these changes. Why do you want to get another default plan. What is with the current default plan?

File: openslides_votecollector/views.py [hunk 9]
```diff
@@ -328,12 +345,8 @@ def on_start(self, item):
...
     def on_start(self, obj):
         # Clear in_range and battery_level of all keypads.
-        # Attention: Cannot use Keypad.objects.all().update(in_range=False, battery_level=-1)
-        # since no post_save signals will be sent on update.
-        for keypad in Keypad.objects.all():
-            keypad.in_range = False
-            keypad.battery_level = -1
-            keypad.save()
+        Keypad.objects.all().update(in_range=False, battery_level=-1)
+        # inform_changed_data(Keypad.objects.all())
```
Comment (normanjaeckel) [inline openslides_votecollector/views.py]: What is with this autoupdate here?

File: openslides_votecollector/views.py [hunk 10]
```diff
@@ -437,18 +457,92 @@ def post(self, request, poll_id, keypad_id):
...
+        # Bulk create ballots and connections.
+        if ballot:
+            ballot.save_ballots()
+        conn_model.objects.bulk_create(connections)
+
+        # Trigger auto update.
+        connections = conn_model.objects.filter(poll=poll, keypad_id__in=keypad_set)
+        inform_changed_data(connections)
+
+        return HttpResponse('')
```
Comment (normanjaeckel) [inline openslides_votecollector/views.py]: Do you want to provide a success message?

File: openslides_votecollector/views.py [hunk 11]
```diff
@@ -463,15 +557,23 @@ def post(self, request, poll_id, keypad_id):
...
+                    conn = MotionPollKeypadConnection.objects.get(poll=poll, keypad=keypad)
+                except MotionPollKeypadConnection.DoesNotExist:
+                    conn = MotionPollKeypadConnection()
+                    conn.poll = poll
+                    conn.keypad = keypad
+                conn.serial_number = request.POST.get('sn')
+                conn.value = value
+                conn.save()
+            else:
+                return HttpResponse(_('Vote rejected'))
```
Comment (normanjaeckel) [inline openslides_votecollector/views.py]: Please try not to return anything deep inside the function. The code would be much more readable.

File: openslides_votecollector/views.py [hunk 12]
```diff
@@ -491,14 +593,82 @@ def post(self, request, poll_id, keypad_id):
...
+class Candidates(utils_views.View):
+    http_method_names = ['post']
+
+    @transaction.atomic()
+    def post(self, request, poll_id):
+        # Get assignment poll.
+        try:
+            poll = AssignmentPoll.objects.get(id=poll_id)
+        except AssignmentPoll.DoesNotExist:
+            return HttpResponse('')
```
Comment (normanjaeckel) [inline openslides_votecollector/views.py]: Here the same.

File: openslides_votecollector/views.py [hunk 13]
```diff
@@ -491,14 +593,82 @@ def post(self, request, poll_id, keypad_id):
...
+            keypad_set.add(keypad.id)
+
+        # Bulk create connections.
+        AssignmentPollKeypadConnection.objects.bulk_create(connections)
+
+        # Trigger auto update.
+        connections = AssignmentPollKeypadConnection.objects.filter(poll=poll, keypad_id__in=keypad_set)
+        inform_changed_data(connections)
+
+        return HttpResponse('')
```
Comment (normanjaeckel) [inline openslides_votecollector/views.py]: Here the same.

File: openslides_votecollector/views.py [hunk 14]
```diff
@@ -576,7 +747,36 @@ def post(self, request, item_id, keypad_id):
...
+            # Mark keypad as in range and update battery level.
+            keypad.in_range = True
+            keypad.battery_level = vote['bl']
+            keypad.save(skip_autoupdate=True)
+            keypads.append(keypad)
+
+        # Trigger auto-update.
+        inform_changed_data(keypads)
+
+        return HttpResponse('')
```
Comment (normanjaeckel) [inline openslides_votecollector/views.py]: Here the same.

File: openslides_votecollector/api.py (hunk 1 above)
Comment (jwinzer) [inline openslides_votecollector/api.py]: Accepted

File: openslides_votecollector/api.py (hunk 2 above)
Comment (jwinzer) [inline openslides_votecollector/api.py]: This was temporarily commented out and will be corrected.

File: openslides_votecollector/api.py (hunk 3 above)
Comment (jwinzer) [inline openslides_votecollector/api.py]: Accepted

File: openslides_votecollector/api.py (hunk 4 above)
Comment (jwinzer) [inline openslides_votecollector/api.py]: These can be network errors and others, usually TimeoutError. So for now I would like to keep this generic.

File: openslides_votecollector/seating_plan.py (hunk 8 above)
Comment (jwinzer) [inline openslides_votecollector/seating_plan.py]: Oh, this is a seating plan for a specific customer. I will restore the original plan.

File: openslides_votecollector/views.py (hunk 9 above)
Comment (jwinzer) [inline openslides_votecollector/views.py]: Autoupdate not wanted here. I will add a comment.

File: openslides_votecollector/views.py (hunk 10 above)
Comment (jwinzer) [inline openslides_votecollector/views.py]: No, not really.

File: openslides_votecollector/api.py (hunk 4 above)
Comment (jwinzer) [inline openslides_votecollector/api.py]: Let's mark this for future review. In current use cases I don't expect any other errors except for a Timeout due to the fact that VC is not running.

**Comment by emanuelschuetze** [comment]
See #35.

**Comment by emanuelschuetze** [comment]
Sorry, I closed the wrong pr! #34 is outdated and will be closed. This pr (#35) should be merged...

**Comment by emanuelschuetze** [comment]
@normanjaeckel Please review the last commit asap. I have to merge it today.

**Comment by normanjaeckel** [comment]
I still strongly recommend not to use a global except statement. At the moment the api.py is even inconsistent for it uses except Exception once and only except in all other cases. Please open at least an issue to provide explicit expected errors.

**Comment by emanuelschuetze** [comment]
Issue #37 created. Now I'll rebase...

**Comment by emanuelschuetze** [comment]
Rebased.

**Pull request merged**